Tests/MessagingPush/PushHandling/AutomaticPushClickHandlingIntegrationTests.swift

```diff
@@ -0,0 +1,214 @@
+     ... this call to the host app's current delegate means that this function is called again. Once this function is called again, we have gotten into an infinite loop.
+     */
+
+    func test_givenMultiplePushClickHandlers_expectOtherClickHandlerHandlesPush_expectInfiniteLoopPrevented() {
+        setupAutomaticPushClickHandlingFeature()
+
+        notificationCenterMock.currentDelegate = UNUserNotificationCenterDelegateMock()
+        otherPushClickDelegate.userNotificationCenterDidReceiveCalled = { _ in
+            // handle push click
+
+            // What other SDKs (such as FCM SDK) may do in it's swizzling implementation.
+            // Is to make a call to the host app's current UserNotificationCenter.delegate (which is our SDK).
+            // This line of code is what could cause an infinite loop unless we prevent it.
+            clickOnPushNotification(givenPush)
+        }
+
+        let givenPush = getPush(content: [
+            "CIO": [
+                "link": "TODO"
+            ]
+        ])
+
+        clickOnPushNotification(givenPush)
+
+        XCTAssertEqusl(pushClickHandlerMock.pushClickedCount, 1) // should only be handled 1 time
+    }
+
+    // MARK: local push notification support
+
+    func test_givenMultiplePushClickHandlers_expectOtherClickHandlerHandlesPush_expectInfiniteLoopPrevented() {
+        setupAutomaticPushClickHandlingFeature()
+
+        notificationCenterMock.currentDelegate = UNUserNotificationCenterDelegateMock()
+        otherPushClickDelegate.userNotificationCenterDidReceiveCalled = { _ in
+            // handle push click
+        }
+
+        let givenLocalNotification = getLocalNotification(pushId: "hard-coded-non-unique-id")
+
+        clickOnPushNotification(givenLocalNotification)
+
+        XCTAssertEqusl(pushClickHandlerMock.pushClickedCount, 1) // should only be handled 1 time
+
+        clickOnPushNotification(givenLocalNotification)
+
+        XCTAssertEqusl(pushClickHandlerMock.pushClickedCount, 2) // should be handled again, even though we have already handled push with that push ID before.
+    }
+    
+    // TODO: the test class here tests only when a push is clicked. Add test functions for when a push is received on the device and we need to determine if we should show the push while app in foreground or not. 
+    // Another way to put it, we need to test our SDK's logic of: 
+    // userNotificationCenter(willPresent:)
+}
+
+extension AutomaticPushClickHandlingInsetegrationTest {
+    func setupAutomaticPushClickHandlingFeature() {
+        // TODO: we might need to change this function body because we may not be able to run the swizzling code in the test function. Running this line below will setup the swizzling code.
```

Comment:
Bringing attention to this important note. 

I have not yet been able to create automated tests around Swift swizzling that we can rely on. 

Therefore, expect in a future PR that our automated tests do not actually executed the swizzling code in our SDK for this feature. Instead, we will *mock/stub* the swizzling code by instead directly calling the functions in our SDK that the swizzling code calls. This means that we will still be able to have automated tests around the logic of this feature, but some parts of this feature (such as swizzling) will still need to be tested manually. 

In the future, we can decide how we test swizzling code.